include/libcyphal/transport/udp/udp_transport_impl.hpp

```diff
@@ -0,0 +1,708 @@
+        const RequestTxParams& params) override
+    {
+        auto any_error = ensureMediaTxSockets();
+        if (any_error.has_value())
+        {
+            return any_error.value();
+        }
+
+        return SvcRequestTxSession::make(memoryResources().general, asDelegate(), params);
+    }
+
+    CETL_NODISCARD Expected<UniquePtr<IResponseRxSession>, AnyError> makeResponseRxSession(
+        const ResponseRxParams& params) override
+    {
+        // TODO: Uncomment!
+        //        const cetl::optional<AnyError> any_error = ensureNewSessionFor(CanardTransferKindResponse,
+        //        params.service_id); if (any_error.has_value())
+        //        {
+        //            return any_error.value();
+        //        }
+
+        return SvcResponseRxSession::make(memoryResources().general, asDelegate(), params);
+    }
+
+    CETL_NODISCARD Expected<UniquePtr<IResponseTxSession>, AnyError> makeResponseTxSession(
+        const ResponseTxParams& params) override
+    {
+        auto any_error = ensureMediaTxSockets();
+        if (any_error.has_value())
+        {
+            return any_error.value();
+        }
+
+        return SvcResponseTxSession::make(memoryResources().general, asDelegate(), params);
+    }
+
+    // MARK: IRunnable
+
+    CETL_NODISCARD IRunnable::MaybeError run(const TimePoint now) override
+    {
+        cetl::optional<AnyError> any_error{};
+
+        // We deliberately first run TX as much as possible, and only then running RX -
+        // transmission will release resources (like TX queue items) and make room for new incoming frames.
+        //
+        any_error = runMediaTransmit(now);
+        if (any_error.has_value())
+        {
+            return any_error.value();
+        }
+
+        return {};
+    }
+
+    // MARK: TransportDelegate
+
+    CETL_NODISCARD TransportDelegate& asDelegate()
+    {
+        return *this;
+    }
+
+    CETL_NODISCARD cetl::optional<AnyError> sendAnyTransfer(const AnyUdpardTxMetadata::Variant& tx_metadata_var,
+                                                            const PayloadFragments payload_fragments) override
+    {
+        // Udpard currently does not support fragmented payloads (at `udpardTx[Publish|Request|Respond]`).
+        // so we need to concatenate them when there are more than one non-empty fragment.
+        // TODO: Make similar issue but for Udpard repo.
+        // See https://github.com/OpenCyphal/libcanard/issues/223
+        //
+        const ContiguousPayload payload{memoryResources().general, payload_fragments};
+        if ((payload.data() == nullptr) && (payload.size() > 0))
+        {
+            return MemoryError{};
+        }
+
+        cetl::optional<AnyError> opt_any_error;
+
+        for (Media& some_media : media_array_)
+        {
+            opt_any_error = withEnsureMediaTxSocket(some_media,
+                                                    [this, &tx_metadata_var, &payload](auto& media, auto& tx_socket)
```

Comment:
This lambda is probably going to exceed most small-function optimizations and end up calling malloc. Can we make this a function on the callstack that this lambda delegates to? Nobody is storing this lambda after the callstack unwinds, right?

Reply:
This code does not make use of any type-erasing function containers, so the matters of small-size optimizations are not applicable here. There is nothing that could use dynamic memory here.

Reply:
It's unfair of me to ask this without an ability to test so I'll retract this comment for now but we may have to revisit this in the future.

We have seen that there is a threshold of complexity or perhaps some way of referencing the lambda that causes an `std::function` to be generated instead. `std::function` does have a small-function optimization that will fall-back to malloc.

Found this with some googling: https://stackoverflow.com/questions/70590194/small-object-optimization-useless-in-using-stdfunction

Reply:
I would like to learn more about this. The referenced StackOverflow post does not seem to be related (unless I'm misreading it) because it takes about std::function specifically, which we do not use. Are you saying a plain lambda can be boxed into an std::function? That sounds wild.

Reply:
It's that lambdas can only be converted into function pointers if they do not capture. Once you capture the closure will have to be stored somewhere. This means that, when passing a lambda as an argument, any overrides that take `std::function` will be selected which may or may not be able to store the capture in the implementation's small-object buffer.

Reply:
I understand that; it's the documented behavior of std::function. We are not using std::function so it would seem that this is a non-issue, or am I missing something?

Reply:
We forward the lambda directly without any type-erasure.